ledger/participant-integration-api/src/main/scala/platform/index/BuffersUpdater.scala

```diff
@@ -0,0 +1,168 @@
+
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.stream.scaladsl.{Keep, RestartSource, Sink, Source}
+import akka.stream._
+import akka.{Done, NotUsed}
+import com.daml.ledger.participant.state.v1.Offset
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.index.BuffersUpdater._
+import com.daml.platform.store.appendonlydao.EventSequentialId
+import com.daml.platform.store.appendonlydao.events.{Contract, Key, Party}
+import com.daml.platform.store.dao.events.ContractStateEvent
+import com.daml.platform.store.interfaces.TransactionLogUpdate
+import com.daml.scalautil.Statement.discard
+
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.{Failure, Success}
+import scala.util.control.NonFatal
+
+/** Creates and manages a subscription to a transaction log updates source
+  * (see [[com.daml.platform.store.dao.LedgerDaoTransactionsReader.getTransactionLogUpdates]]
+  * and uses for updating:
+  *    * The transactions buffer used for in-memory Ledger API serving.
+  *    * The mutable contract state cache.
+  *
+  * @param subscribeToTransactionLogUpdates Subscribe to the transaction log updates stream starting at a specific
+  *                                         `(Offset, EventSequentialId)`.
+  * @param updateTransactionsBuffer Trigger externally the update of the transactions buffer.
+  * @param toContractStateEvents Converts [[TransactionLogUpdate]]s to [[ContractStateEvent]]s.
+  * @param updateMutableCache Trigger externally the update of the mutable contract state cache.
+  *                           (see [[com.daml.platform.store.cache.MutableCacheBackedContractStore]])
+  * @param minBackoffStreamRestart Minimum back-off before restarting the transaction log updates stream.
+  * @param mat The Akka materializer.
+  * @param loggingContext The logging context.
+  * @param executionContext The execution context.
+  */
+private[index] class BuffersUpdater(
+    subscribeToTransactionLogUpdates: SubscribeToTransactionLogUpdates,
+    updateTransactionsBuffer: (Offset, TransactionLogUpdate) => Unit,
+    toContractStateEvents: TransactionLogUpdate => Iterator[ContractStateEvent],
+    updateMutableCache: ContractStateEvent => Unit,
+    minBackoffStreamRestart: FiniteDuration,
+    sysExitWithCode: Int => Unit,
+)(implicit mat: Materializer, loggingContext: LoggingContext, executionContext: ExecutionContext)
+    extends AutoCloseable {
+
+  private val logger = ContextualizedLogger.get(getClass)
+
+  private[index] val updaterIndex = new AtomicReference(
+    Offset.beforeBegin -> EventSequentialId.beforeBegin
+  )
+
+  private val (transactionLogUpdatesKillSwitch, transactionLogUpdatesDone) =
+    RestartSource
+      .withBackoff(
+        RestartSettings(
+          minBackoff = minBackoffStreamRestart,
+          maxBackoff = 10.seconds,
+          randomFactor = 0.2,
```

Comment:
randomFactor helps when there are multiple consumers hitting the resource at the same time. I don't think it is the case here.

Reply:
True. Changed to `0.0`